### PR TITLE
[WIP] feat(core): Add support for extension point/extension pattern on top of Context

### PR DESCRIPTION
**Pull request by raymondfeng**
Extension point/extension is a pattern for extensibility.
This PR illustrates how we can support it using the Context apis
and naming conventions. It also serves as an invitation to discuss
if we should support such entities out of box.

#### Related issues

- connect to https://github.com/strongloop/loopback-next/issues/543

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: packages/context/test/acceptance/extension-point.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    const key = `${this.name}.${name}`;
+    const extensions = this.getExtensionBindingMap();
+    return extensions[key];
+  }
+
+  /**
+   * Get an instance of an extension by name
+   * @param name Name of the extension
+   */
+  async getExtension(name: string): Promise<EXT> {
```
Comment (virkt25) [inline packages/context/test/acceptance/extension-point.ts]: Would it make sense for us to automatically return the extension if only one has been registered to an ExtensionPoint? --> or mark a default one to use? ... I think it makes for an better developer experience.

As in the example below -> A developer with only one AuthStrategy would find it tedious to have to specify `local` again and again ...

File: packages/context/test/acceptance/extension-point.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        },
+      },
+    });
+
+    // Register multiple extensions
+    ctx
+      .bind('authentication.strategies.local')
+      .toClass(LocalStrategy)
+      .inScope(BindingScope.SINGLETON)
+      .tag('authentication-strategy');
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: There are a few ways to group a set of bindings as extensions for a given extension point.
1. Use naming convention such as `authentication.strategies.*`
2. Use tags (all bindings with tag `'authentication-strategy'` will be treated as extensions for the `authentication.manager` extension point.

File: packages/context/test/acceptance/extension-point.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: When an extension point manages one or more extensions, a few common delegating cases are involved:

1. Pick up one by the matching metadata. For example, an authentication action can use the `authentication.manager` extension point to delegate a corresponding strategy provider based on the decoration of the target controller method.

2. Invoke all extensions, for example, email/sms/push notifications.

3. Try extensions one by one until one of them can handle the request.

4. Only one extension is allowed for a given type. For example, Password hashing function.

File: packages/context/test/acceptance/extension-point.feature.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+```ts
+class LocalStrategy implements AuthenticationStrategy {
+  constructor(@inject('config') private config: LocalConfig) {}
+}
+```
+
+# TBD
+
+- Should we go beyond naming conventions to make ExtensionPoint/Extension first class entity of the context?
+- Should we introduce decorators such as `@extensionPoint` and `@extension`?
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.feature.md]: `@extension` decorator can be used to denote the `name` of an extension point it contributes to.

**Comment by raymondfeng** [comment]
Yes, I'm advocating to come up an out-of-box solution for extension point/extension pattern so that we can:

1. Allow multiple servers for the REST component
2. Allow multiple strategies for authentication
3. Allow multiple logging implementations
4. Allow multiple deserialization/serialization schemes for different media types
...

File: packages/context/test/acceptance/extension-point.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+     */
+    configuration?: ExtensionPointConfig,
+  ) {
+    this.configuration = configuration || {extensionPoint: {}, extensions: {}};
+  }
+
+  /**
+   * Find an array of bindings for extensions
+   */
+  getExtensionBindings(): Binding[] {
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: The difference between names `getExtensionBindings` and `getExtensionBinding` (below) is too small and easy to overlook. I am proposing to rename this one to `getAllExtensionBindings`.

File: packages/context/test/acceptance/extension-point.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  /**
+   * Get an instance of an extension by name
+   * @param name Name of the extension
+   */
+  async getExtension(name: string): Promise<EXT> {
+    const binding = this.getExtensionBinding(name);
+    // Create a child context to bind `config`
+    const extensionContext = new Context(this.context);
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: We need to carefully consider implications for CONTEXT-scoped bindings, as they will be cached at the level of this temporary extension context. Is this the right thing to do? Shouldn't we cache CONTEXT-scoped bindings at `this.context` level instead?

File: packages/context/test/acceptance/extension-point.feature.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    ldap: { /* config for ldap strategy */}
+    oauth2: { /* config for oauth2 strategy */}
+  }
+}
+```
+- Each of the strategy class should be able to receive its configuration via dependency injection. For example,
+
+```ts
+class LocalStrategy implements AuthenticationStrategy {
+  constructor(@inject('config') private config: LocalConfig) {}
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.feature.md]: +1, this is great.

File: packages/context/test/acceptance/extension-point.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class AuthenticationManager extends ExtensionPoint<AuthenticationStrategy> {
+    constructor(
+      @inject('$context') context: Context,
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: I am not very happy that we need to inject the full context object here, but I am not sure if there is anything we can do about that.

File: packages/context/test/acceptance/extension-point.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+  class OAuth2Strategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class AuthenticationManager extends ExtensionPoint<AuthenticationStrategy> {
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: Using the name`Manager` is considered as antipattern, see e.g [here](https://stackoverflow.com/questions/1866794/naming-classes-how-to-avoid-calling-everything-a-whatevermanager) and [here](https://stackoverflow.com/questions/5569666/what-is-a-better-name-than-manager-processor-etc).

Let's use a more descriptive name please, for example `AuthenticationHandler`.

File: packages/context/test/acceptance/extension-point.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply packages/context/test/acceptance/extension-point.ts]: > Use tags (all bindings with tag 'authentication-strategy' will be treated as extensions for the authentication.manager extension point.

I think it's a bad idea to have two possibly conflicting conventions for looking up extensions. For example, it allows you to bind authentication strategy as a controller too via  `ctx.bind('controllers.authentication').toClass(MyStrategy).tag('authentication-strategy')`. I know this example is silly, but my concern is that users often do silly things and then we have to spend our precious time explaining them what they did wrong.

I personally prefer to get rid of Binding's `.tag()` method entirely, considering that we haven't used it anywhere so far, and we are already almost in Beta.

Alternatively, if you insist on supporting both naming convention and binding tags to look up extensions, then please capture this in acceptance tests, including how edge cases like my silly example above should be handled.

File: packages/context/test/acceptance/extension-point.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+// tslint:disable:no-any
+/**
+ * Interface for the extension point configuration
+ */
+export interface ExtensionPointConfig {
+  extensions: {
+    [extensionName: string]: any;
+  };
+  extensionPoint: {
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: What is the point of this member? AFAICT, it's not used anywhere else in this pull request.

File: packages/context/test/acceptance/extension-point.ts [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+  });
+
+  function createContext() {
+    ctx = new Context();
+
+    ctx.bind('$context').to(ctx);
+    // Register the extension point
+    ctx.bind('authentication.manager').toClass(AuthenticationManager);
+    ctx.bind('authentication.config').to({
+      extensions: {
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: Can we avoid this extra level of nesting (config -> extensions -> local) and go for shorter config -> local?

File: packages/context/test/acceptance/extension-point.ts [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        oauth2: {
+          clientId: 'my-client-id',
+          clientSecret: 'my-client-secret',
+          tokenInspectUrl: 'https://localhost:3000/oauth2/inspect',
+        },
+      },
+    });
+
+    // Register multiple extensions
+    ctx
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: I'd like to see some kind of helpers to simplify the registration of extensions. Ideally, I'd want the registration code to refer to extension point by constructor (`AuthenticationHandler`) and then have everything else filled for me by the framework.

For example:

```ts
app.bindExtensionOf(AuthenticationHandler).toClass(LocalStrategy);
```

(I am using `AuthenticationHandler` instead of `AuthenticationManager` per my earlier comment.)

File: packages/context/test/acceptance/extension-point.ts [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  /**
+   * Look up an extension binding by name
+   * @param name Name of the extension
+   */
+  getExtensionBinding(name: string): Binding {
+    const key = `${this.name}.${name}`;
+    const extensions = this.getExtensionBindingMap();
+    return extensions[key];
```
Comment (bajtos) [inline packages/context/test/acceptance/extension-point.ts]: I am proposing to simplify this method to the following:

```ts
return this.context.getBinding(key);
```

The current implementation builds a full map of all extension bindings, which is a preliminary pessimization IMO.

There is also a subtle difference handling of unknown extension names: the current proposal returns undefined, my proposal throws error `The key ${key} was not bound to any value.`.

File: packages/context/test/acceptance/extension-point.feature.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+}
+```
+
+# TBD
+
+- Should we go beyond naming conventions to make ExtensionPoint/Extension first class entity of the context?
```
[the remark this answers is not in the document]
Comment (bajtos) [reply packages/context/test/acceptance/extension-point.feature.md]: I agree that we should provide interface/base classes OOTB. However, I think they should be part of a different package than `@loopback/context`, probably `@loopback/core`.

[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.feature.md]: I'll move it into `@loopback/core`. This way, `@loopback/context` stays as the IoC container while core implements common patterns on top of `Context`, for example, `application`, `extension point/extension`, and `component`.

File: packages/context/test/acceptance/extension-point.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: I'm not proposing to support both. Personally, I don't want to rely too much on the naming magic and prefer to use `tag` to explicitly group bound services. Tags can also be used as filters for `@inject`. If we visualize the `Context`, tags will be really useful too.

File: packages/context/test/acceptance/extension-point.ts (hunk 9 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: I was trying to separate config properties for the extension point itself from the ones for extensions.

File: packages/context/test/acceptance/extension-point.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: +1.

File: packages/context/test/acceptance/extension-point.ts (hunk 5 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: Let me think about that.

Meanwhile, I'll evaluate the possibility to make `ExtensionPoint` extend `Context`. This way, each extension point will have its own isolated context.

File: packages/context/test/acceptance/extension-point.ts (hunk 11 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/extension-point.ts]: +1.

File: packages/core/test/acceptance/extension-point.ts [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+  constructor(@inject('config') public config: any) {}
+
+  authenticate(credentials: any) {
+    return Promise.resolve(true);
+  }
+}
+
+const AUTH_EXTENSION_POINT = 'authentication.strategies';
+
+class AuthenticationManager extends ExtensionPoint<AuthenticationStrategy> {
```
Comment (bajtos) [inline packages/core/test/acceptance/extension-point.ts]: re-posting https://github.com/strongloop/loopback-next/pull/657#discussion_r145364249

Using the name`Manager` is considered as antipattern, see e.g [here](https://stackoverflow.com/questions/1866794/naming-classes-how-to-avoid-calling-everything-a-whatevermanager) and [here](https://stackoverflow.com/questions/5569666/what-is-a-better-name-than-manager-processor-etc).

Let's use a more descriptive name please, for example `AuthenticationHandler`.

On a side note, I find it pretty annoying when my review comments are ignored without any explanation 😞

**Comment by bajtos** [comment]
This patch adds quite a few new concepts, please give me few more days to digest it, so that I can provide meaningful feedback and review.

**Comment by bajtos** [comment]
@raymondfeng could you please update our reference implementation (see https://github.com/strongloop/loopback-next/tree/master/packages/authentication) to use this new Extension/ExtensionPoint API too?

I'd like us to consider the impact of this proposal on multiple server instances/configs, is there a way how to have these two use cases following the same design convention? See https://github.com/strongloop/loopback-next/pull/671#issuecomment-338691103 for more details.

I am still concerned about https://github.com/strongloop/loopback-next/pull/657#discussion_r145362995, I think it can have nasty impact - let me try to come up with an example to illustrate my concerns.
> We need to carefully consider implications for CONTEXT-scoped bindings, as they will be cached at the level of this temporary extension context. Is this the right thing to do? Shouldn't we cache CONTEXT-scoped bindings at this.context level instead?

**Comment by raymondfeng** [comment]
@slnode test please

**Comment by bajtos** [comment]
There have been no updates in this pull request for more than a year by now, I am closing it for now. @raymondfeng feel free to reopen if you get time to continue this work. (Considering how outdated these changes are, it may be better to start from scratch.)
